desc/singularities.py

```diff
@@ -0,0 +1,758 @@
+            src_data_polar,
+            diag=True,
+        ).reshape((eval_grid.num_nodes, kernel.ndim))
+
+        dS = (v[i] * src_data_polar["|e_theta x e_zeta|"])[:, None]
+        fi = k * dS
+        return fi
+
+    def polar_pt_loop(i, f):
+        # this calculates the effect at a single evaluation point, from all others
+        # in a single field period. loop this to get all pts
+        f_temp = polar_pt_vmap(i)
+        return f + f_temp.reshape((eval_grid.num_nodes, kernel.ndim))
+
+    f = jnp.zeros((eval_grid.num_nodes, kernel.ndim))
+    # vmap found more efficient than fori_loop, esp on gpu, but uses more memory
+    if loop:
+        f = fori_loop(0, v.size, polar_pt_loop, f)
+    else:
+        f = vmap(polar_pt_vmap)(jnp.arange(v.size)).sum(axis=0)
+
+    # we sum distance vectors, so they need to be in xyz for that to work
+    # but then need to convert vectors back to rpz
+    if kernel.ndim == 3:
+        f = xyz2rpz_vec(f, phi=eval_data["phi"])
+
+    return f
+
+
+def singular_integral(
+    eval_data,
+    src_data,
+    kernel,
+    interpolator,
+    loop=False,
+):
+    """Evaluate a singular integral transform on a surface.
+
+    eg f(θ, ζ) = ∫ ∫ K(θ, ζ, θ', ζ') g(θ', ζ') dθ' dζ'
+
+    Where K(θ, ζ, θ', ζ') is the (singular) kernel and g(θ', ζ') is the metric on the
+    surface. See eq. 3.7 in [1]_, but we have absorbed the density σ into K
+
+    Uses method by Malhotra et. al. [1]_ [2]_
+
+    Parameters
+    ----------
+    eval_data : dict
+        Dictionary of data at evaluation points (eval_grid passed to interpolator).
+        Keys should be those required by kernel as kernel.keys. Vector data should be
+        in rpz basis.
+    src_data : dict
+        Dictionary of data at source points (src_grid passed to interpolator). Keys
+        should be those required by kernel as kernel.keys. Vector data should be in
+        rpz basis.
+    kernel : str or callable
+        Kernel function to evaluate. If str, one of the following:
+            '1_over_r' : 1 / |𝐫 − 𝐫'|
+            'nr_over_r3' : 𝐧⋅(𝐫 − 𝐫') / |𝐫 − 𝐫'|³
+            'biot_savart' : μ₀/4π 𝐊×(𝐫 − 𝐫') / |𝐫 − 𝐫'|³
+            'biot_savart_A' : μ₀/4π 𝐊 / |𝐫 − 𝐫'|
+        If callable, should take 3 arguments:
+            eval_data : dict of data at evaluation points (primed)
+            src_data : dict of data at source points (unprimed)
+            diag : boolean, whether to evaluate full cross interactions or just diagonal
+        If a callable, should also have the attributes ``ndim`` and ``keys`` defined.
+        ``ndim`` is an integer representing the dimensionality of the output function f,
+        1 if f is scalar, 3 if f is a vector, etc.
+        ``keys`` is a list of strings of what data is required to evaluate the kernel.
+        The kernel will be called with dictionaries containing this data at source and
+        evaluation points.
+        If vector valued, the input to the kernel function will be in rpz and output
```

Comment:
this is true even if it is not vector valued (like the `1/r` kernel you have), right? the src_data needs to be `rpz`?

Reply:
I guess technically, but if there arent't any vectors than it doesnt matter. The eval/src pts themselves are always built from 1d arrays R, phi, Z